### ISSUE TITLE
Profiling discussion/improvements

### ISSUE THREAD

**Issue by sigbjobo** [sigbjobo/hPF_MD_PMESH#3]
New benchmarks are reported in the benchmark folder, with profiling of small and large systems for up to thousands of cores. 

Current challenges are: 
1. getting the code to scale better for 10 million particles.
2. Obtaining as consistent as possible maximum speed (steps/seconds) for any system size by increasing number of cpus.
3. Determining whether domain-decomposition by layout.exchange can speed up simulations.

**Comment by rainwoodman** [comment]
pm:make_duo is a huge offender at:

https://raw.githubusercontent.com/sigbjobo/hPF_MD_PMESH/MPI/benchmarks/benchmark_1E8/cpu_4096.png

It appears to get worse with more CPUs. I'll investigate this further this week.

**Comment by rainwoodman** [comment]
make_duo is a relative expensive function (making FFT plans), but it is called only per ParticleMesh object.
The number of calls is very low (4) in the 4096 run. The fraction is large only because there were fewer time steps (21 calls to UPDATE FIELD).

**Comment by rainwoodman** [comment]
Also worth noting that in the 256 core run benchmark, the Python startup (find_and_load modules) showed up as a main time consumer:
https://raw.githubusercontent.com/sigbjobo/hPF_MD_PMESH/MPI/benchmarks/benchmark_1E6/cpu_256.png

Did you look into ways of compiling a python script to a standalone application? Then slurm takes care of deploying the standalone binary to the computing nodes, and avoid the startup cost due to jamming the central file system.

There were a few ways of doing this. A relatively new technology recently popped up is
https://gregoryszorc.com/blog/2019/06/24/building-standalone-python-applications-with-pyoxidizer/
(but it is built with rust...)

**Comment by sigbjobo** [comment]
> make_duo is a relative expensive function (making FFT plans), but it is called only per ParticleMesh object.
> The number of calls is very low (4) in the 4096 run. The fraction is large only because there were fewer time steps (21 calls to UPDATE FIELD).

Agree, this is just present at the initialization. Please note that the benchmark plots does not include the initialization, and that for the 100 million particles [speedup plot](https://github.com/sigbjobo/hPF_MD_PMESH/blob/MPI/benchmarks/benchmark_1E8/speedup.png) the scaling is not exactly where we would like to be. I strongly suspect that the levelling off is due too much communication and not having domain decomposition. In other words, I think that we need to figure out the domain decomposition issue (which I think is due to some simple misunderstanding by me) to reduce the amount of communication to show the true power of the pmesh code :) 

Also, if you prefer I can make the flowchart from only the simulation steps, however I find it nice also to know whether there are problems in the scalability of starting the program.

**Comment by sigbjobo** [comment]
> Also worth noting that in the 256 core run benchmark, the Python startup (find_and_load modules) showed up as a main time consumer:
> https://raw.githubusercontent.com/sigbjobo/hPF_MD_PMESH/MPI/benchmarks/benchmark_1E6/cpu_256.png
> 
> Did you look into ways of compiling a python script to a standalone application? Then slurm takes care of deploying the standalone binary to the computing nodes, and avoid the startup cost due to jamming the central file system.
> 
> There were a few ways of doing this. A relatively new technology recently popped up is
> https://gregoryszorc.com/blog/2019/06/24/building-standalone-python-applications-with-pyoxidizer/
> (but it is built with rust...)

This looks like extremely nice solution, I will try it out!

**Comment by sigbjobo** [comment]
I have made a new generation of benchmarks (benchmark_v2 files).

First, my impressions is that the heavy particle-mesh computations are scaling excellently, especially the Fourier transform which is cut by half every time we double the processors.

What remains to be explored is the limits to which we can reduce communication for large systems and many cpus (see [profile digram for largest simulation](https://github.com/sigbjobo/hPF_MD_PMESH/blob/MPI/benchmarks/benchmark_1E8_v2/cpu_8192.png) )

In experimenting casually, the best performance I got was by exchanging particles only the first step, and not inside the simulation loop. This suggests that exchange is probably not needed as frequently, if the particles do not move far.

Any ideas?

**Comment by sigbjobo** [comment]
To me it looks like that domain_decomposition should be done less frequently, and that adding smoothing might be smart to reduce the exchange that is present in the readout used for the force computation.

**Comment by rainwoodman** [comment]
Yes. Reducing the frequency of domain_decomposition is often used as a way
to reduce the communication cost, in the context of a hybrid PM + expensive
shortrange force
calculation.

However I am not so sure how it works out in our case?

We have to create ghost particles during paint / readout if a particle has
moved outside the local PM mesh. The communication cost of that should be
similar to the domain decomposition, I would expect?

The best way to settle this is to measure the cost. For example, we can
compute at the number of particles exchanged that needs to be exchanged:

commcost = comm.allreduce(
  sum([layout.sendcount[i] for i in range(comm.size) if i != comm.rank])
)
(also see
https://github.com/rainwoodman/pmesh/blob/master/pmesh/domain.py#L86)

We can measure this before the exchange in domain decomposition, and also
measure this after the paint/readout layout is created.

The total communication cost would be the sum of all commcost (perhaps
scale by the number of bytes exchanged per particle). Then we can find a
way to justify when to do a domain decomposition, which heuristically
minimizes the communication cost?

(Also, be aware that a non-trivial fraction of time in domain_decomposition
is probably in the calculation of the layout object.)

Yu

**Comment by sigbjobo** [comment]
I will try computing this number asap, should tell us how much communication is reduced upon exchange.

One more thing, is the problems with communication we are facing somehow due too many grid points compared to particles, or using PMESH with unusual setting?

**Comment by rainwoodman** [comment]
The FFT mesh caused communication is accounted in FFT (r2c c2r). Therefore,
the time you saw on profiling shouldn't be caused by the size of FFT mesh.

**Comment by sigbjobo** [comment]
> The FFT mesh caused communication is accounted in FFT (r2c c2r). Therefore, the time you saw on profiling shouldn't be caused by the size of FFT mesh.


That is exactly what I thought too. FFT works superbly in all the scaling tests.
